core/src/main/java/jenkins/security/QueueItemAuthenticatorMonitor.java

```diff
@@ -136,12 +150,100 @@ public void onLeft(Queue.LeftItem li) {
+                        recordedQueueItems.remove(wrappedQueueId); // equals/hashCode only look at queueId, so this works
+                        recordedQueueItems.add(wrappedQueueId);
+                        while (recordedQueueItems.size() > 10) {
+                            // limit to 10 items
+                            recordedQueueItems.remove(recordedQueueItems.first());
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static final class QueueIdAndBuildNumber implements Comparable<QueueIdAndBuildNumber> {
+        long queueId;
+        Long buildNumber;
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            QueueIdAndBuildNumber that = (QueueIdAndBuildNumber) o;
+            return queueId == that.queueId;
+        }
+
+        public QueueIdAndBuildNumber(long queueId) {
+            this.queueId = queueId;
+        }
+
+        public QueueIdAndBuildNumber(long queueId, long buildNumber) {
+            this.queueId = queueId;
+            this.buildNumber = buildNumber;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(queueId);
+        }
+
+        @Override
+        public int compareTo(QueueIdAndBuildNumber o) {
+            return Long.compare(queueId, o.queueId);
+        }
+    }
+
+    @Extension
+    public static class ItemListenerImpl extends ItemListener {
+        @Override
+        public void onDeleted(Item item) {
+            synchronized (buildsLaunchedAsSystemWithAuthenticatorPresentByJob) {
+                buildsLaunchedAsSystemWithAuthenticatorPresentByJob.remove(item.getFullName());
+            }
+        }
+
+        @Override
+        public void onLocationChanged(Item item, String oldFullName, String newFullName) {
+            synchronized (buildsLaunchedAsSystemWithAuthenticatorPresentByJob) {
+                if (buildsLaunchedAsSystemWithAuthenticatorPresentByJob.containsKey(oldFullName)) {
+                    SortedSet<QueueIdAndBuildNumber> references = buildsLaunchedAsSystemWithAuthenticatorPresentByJob.get(oldFullName);
+                    buildsLaunchedAsSystemWithAuthenticatorPresentByJob.remove(oldFullName);
+                    buildsLaunchedAsSystemWithAuthenticatorPresentByJob.put(newFullName, references);
+                }
             }
         }
     }
 
-    private static boolean anyBuildLaunchedAsSystemWithAuthenticatorPresent = false;
+    /**
+     * Records queue IDs and later build numbers of builds started as SYSTEM if there is an authenticator present.
+     * This is a relatively memory efficient way to get this information.
+     */
+    @Restricted(NoExternalUse.class)
+    public static final Map<String, SortedSet<QueueIdAndBuildNumber>> buildsLaunchedAsSystemWithAuthenticatorPresentByJob = new ConcurrentHashMap<>();
```

Comment:
I'm not so sure about making this public, even if it's restricted. You have all these synchronization concerns around accessing it, so it should really only be accessed via methods and never via the field.

Reply:
It's a concurrent map so I'm not too worries. The synchronization stuff is to make sure the updates requiring multiple statements work as expected.

Reply:
The concurrent map aspect only helps for situations where you're doing something like iterating over the map while it's being updated and not getting an exception. Now the synchronization of this code might not matter that much big picture, though I've worked with so many APIs that had to handle thread safety in the past that I can't help but point out concurrency bugs.